### PR TITLE
[WIP] Check what has happened in past given period of time solves#1421

### PR DESCRIPTION
**Pull request by StlMaris123**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

**Comment by jywarren** [comment]
Cool, good start -- take a look @ujithaperera -- so now, you can begin adding checking what's happened in the given period. Glad to see you're making progress, please keep pushing new changes!

**Comment by ujithaperera** [comment]
This PR is mainly focused on phase 2. Right ?. Yes please keep pushing to this PR then we can discuss and do necessary changes. @StlMaris123 if you have any issues, dont spend too much time. You can share the problem with us. Then we all can tske a look and help you out.

**Comment by StlMaris123** [comment]
Thanks @jywarren and @ujithaperera. Am now trying to figure out how to check the modifications made on a node such as update and comment in the past given period of time. I have not yet figured out how to do it but if there are any suggestions on how to go about it I will greatly appreciate. Thanks.

**Comment by jywarren** [comment]
@publiclab/soc - can folks make some suggestions for @StlMaris123 here? Thanks!

**Comment by jywarren** [comment]
Hi StellaMaris - are you still stuck on this? Can you update us? Thanks!

You can search for Revisions with updated or created dates in a range, or comments similarly. See the Stats controller for examples of usage.

**Comment by StlMaris123** [comment]
Hi Jeff, I ot a quick question, on the nodes schema; we do not have any
parameter  passed in of type datetime. In these parameters for instance
`UPDATE "node" SET "comment" = 0, "title" = 'abc', "changed" = 1502371842,
"created" = 1502371842 WHERE "node"."nid" = 44`  What is the value of that
fixnum on "changed " and "created", and how are we able to find time
difference without datetime? Thank you.

**Comment by jywarren** [comment]
Take a look at the stats controller for some examples, the node timestamps are unfortunately not in standard rails format because they're from when the site was Drupal. Also reach out to @publiclab/soc folks if you need help!

**Comment by StlMaris123** [comment]
I have been trying to follow how time period has been done on the stats controller. Am trying to check when time when a  user subscribed to a node. Their relationship is through tag_selection which does not include time. How can I check time when subscription happened?

File: test/fixtures/node.yml [hunk 1]
```diff
@@ -236,3 +236,25 @@ method:
...
+node2:
+  nid: 22
+  uid: 8
+  title: "Method page"
+  created: <%= Time.now.to_i %>
+  changed: <%= Time.now.to_i %>
+  status: 1
+  type: "page"
+  cached_likes: 0
+  path: "/wiki/spectrometer"
```
Comment (jywarren) [inline test/fixtures/node.yml]: Looks like you have two nodes with the same path and title. Can you choose another example to work from, like `/wiki/kite-mapping`?

File: app/models/user.rb [hunk 2]
```diff
@@ -265,6 +265,11 @@ def questions
     Node.questions.where(status: 1, uid: id)
   end
 
+  def content_followed_in_past_period(time_period)
+     nodes_created = self.node.select {|node| Time.at(node.created) >= time_period }
+     nodes_edited = self.node.select {|node| Time.at(node.changed) >= time_period unless node.changed.empty?}
```
Comment (jywarren) [inline app/models/user.rb]: Here, looks like you may want to use an ActiveRecord query like:

https://github.com/publiclab/plots2/blob/master/app/controllers/stats_controller.rb#L15

```ruby
@notes = Node.select(%i[created type status])
                 .where(type: 'note', status: 1, created: @start.to_i..@end.to_i)
```

**Comment by jywarren** [comment]
Ah, the issue is that anything in quotations won't evaluate -- so you're sending the exact string `"created >= Time.zone.now"` to MySQL -- instead, try:

```ruby
Node.where("created >= #{Time.zone.now}")
```

That way it'll evaluate the Ruby and substitute its output for the section blocked out with `#{...}`. Make sense?

**Comment by StlMaris123** [comment]
Yes. I have tried that too.  What has been interpolated will be executed but comparison wont take place. I get an error like `ActiveRecord::StatementInvalid: SQLite3::SQLException: near "17": syntax error: SELECT "node".* FROM "node"  WHERE (created >= 2017-08-28 17:28:41 UTC)`

**Comment by jywarren** [comment]
ah, ok -- i think you need `Time.zone.now.to_s(:sql)` -- but i'm not sure. I do think that adding `.to_s(:sql)` will format a Time object properly for databases.

**Comment by StlMaris123** [comment]
Thanks Jeff,
That actually gave me a clue

**Comment by jywarren** [comment]
Hmm I see: 

test_should_choose_i18n_for_layout/alerts(HomeControllerTest): ActionView::Template::Error: undefined method `author' for nil:NilClass

app/views/dashboard/_node_wiki.html.erb:6:in `_app_views_dashboard__node_wiki_html_erb__45696825107109115_47096799501960'

**Comment by StlMaris123** [comment]
I'll look at it and try to find out why it affected other tests.

**Comment by StlMaris123** [comment]
@jywarren the tests are failing even without the code I had added

**Comment by jywarren** [comment]
Hmm, I'll see if I can pick at this one in the next few days but things are quite busy over here. @publiclab/soc - do folks have any ideas with this, or could help StellaMaris out?

**Comment by StlMaris123** [comment]
Hi Jeff,
The issue was with the node fixtures

**Comment by ujithaperera** [comment]
Hi @StlMaris123 , It's Great that you already figure out the test failing issue. Please feel free to mention me on comments, specially when you are seeking for a help. Because sometimes I don't get the notifications without a mention.

**Comment by jywarren** [comment]
Oh - v glad you figured it out!

**Comment by jywarren** [comment]
Is this ready to merge then?

**Comment by StlMaris123** [comment]
Yes. Please review. @ujitha. Sorry about that.  I have noted it

**Comment by jywarren** [comment]
@ujitha, do you think this is ready to merge? Thanks!

**Comment by ujithaperera** [comment]
This is good to go @jywarren .  And easy to read and track changes with this size. 
@StlMaris123 your method is fulfilling the basic requirement. Here we have different user account types and different node status values. I think it's good to test for those combinations also before going for live. Add couple of combinations to your test case. that's all.  And also have a look at Jeff's comment over there in diff tab. It's not considered yet, I guess.

**Comment by ujithaperera** [comment]
BTW @StlMaris123 , @jywarren  please use '@ujithaperera' mention tag for me. Actually above mentions ('@ujitha') are not mapped to my account actually.

**Comment by jywarren** [comment]
Oops! I realize i was using your PublicLab.org handle. OK!!! Thanks.

**Comment by jywarren** [comment]
Great, done! Can you now update this and link it from your planning issue? Thanks again!

**Pull request merged**